### DISCUSSION TITLE
useMutations for multiple mutations

### DISCUSSION THREAD

**Discussion by trevoro** [TanStack/query#4009] [Ideas]
Hey,

We have a method for grouping queries `useQueries` but there isn't one for grouping together multiple mutations.
We have some components that contain all the queries and mutations and pass the results down to dumber components and rather than having many variables like `actionFooMutation, actionBazMutation` it would be nice to be able to group them together in a single variable. Something like

```
const mutations = useMutations({
   foo: ['actionFoo', { ...paramsOne }],
   bar: ['actionBar', { ...paramsTwo }],
});

mutations.foo.isLoading
```

etc.

**Comment by TkDodo** [comment]
I think you can build that in user-land:

```
const useMutations = () => {
  const foo = useMutation()
  const bar = useMutation()
  return { foo, bar }
}
```

**Reply by dwjohnston** [reply]
@TkDodo  What if you want to do it with arbitrary numbers (for example, mutation per item in an array).

**Reply by TkDodo** [reply]
you can make one mutation that fires multiple requests

**Comment by dwjohnston** [comment]
I will note that 

```tsx
function useMultipleMutation(ids: Array<string>){
    const ref = useRef<Record<string, any>>({}); 
    ids.forEach((v) => {     
        // eslint-disable-next-line react-hooks/rules-of-hooks
        const result = useMutation(likeTheThing); 
        ref.current[v] = result; 
    })
    return ref; 
} 
```

Although this technically breaks rules of hooks, in my test usage of it, it appears to be not problematic.

**Reply by TkDodo** [reply]
> Although this technically breaks rules of hooks, in my test usage of it, it appears to be not problematic.

it breaks if the length of `ids` is different in each render, because you will call a different amount of hooks then.

**Reply by dwjohnston** [reply]
Yup thanks.

**Reply by akos-sebestyen** [reply]
@TkDodo any update on this in 2024? Looking for a standard way to achieve the same.

**Reply by TkDodo** [reply]
imagine `useMutation` were `useState`. You can't have a variable amount of hooks in the same component. But you can map over an array and create one component per entry, each containing a mutation hook ;)

**Reply by parsadotsh** [reply]
@TkDodo Imagine you want to build a chat app where each new message sent is a mutation. To do optimistic updates, I'm not sure there's a way to do it currently without modifying the query cache (the "UI" approach). Wouldn't a useMutations hook be nice here? Instead of a single "variables" and "isPending" it would return an array of them for all the current active mutations.

**Reply by TkDodo** [reply]
useMutationState can do it.

**Reply by parsadotsh** [reply]
@TkDodo Thank you!! 😄 Works great, I was about to roll my own from scratch haha. 

I now realize this was shown in  "[If the mutation and the query don't live in the same component](https://tanstack.com/query/latest/docs/framework/react/guides/optimistic-updates#if-the-mutation-and-the-query-dont-live-in-the-same-component)" but I didn't really look at it since the title didn't describe my situation. Maybe the title can be amended slightly.

**Reply by Arthur944** [reply]
What if you want to trigger mutations from a central location?

Imagine a table where you can select rows, and press a single button to run an action on each selected row.

Currently the only way I can think of is to do a useMutation in each of the Row components, and put a useEffect in there that triggers a mutations if a prop changes. That feels very wrong.

That or batch all the actions into a single useMutation, but then you'd have to wait for all rows to finish in order to see the results of any one row. If there's one that takes a long time it looks like all the rows are taking forever.

**Reply by TkDodo** [reply]
> Imagine a table where you can select rows, and press a single button to run an action on each selected row.

Trigger the mutation in the table component by calling `.mutate()` once for each row, then have each row track the state with `useMutationState()`
